ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java

```diff
@@ -77,7 +82,7 @@
   private final WindowExpression windowExpression;
   private final List<Expression> aggregateFunctionArguments;
   private final List<FunctionCall> functionList;
-  private final List<Expression> requiredColumnList;
+  private final Set<DereferenceExpression> requiredColumns;
```

Comment:
Actually, I think even if the format was field-order-agnostic, we still use the order to come up with the internal column names. Incidentally, I'm working on expanding what we validate in our compatibility verification tests to include the schemas going in/out of each step, so hopefully this kind of thing gets caught by test failure soon.

Reply:
The order of the columns in the select expression are defined by `finalSelectExpressions` which is still a `List`, so I believe the order is being maintained.

> Incidentally, I'm working on expanding what we validate in our compatibility verification tests to include the schemas going in/out of each step, so hopefully this kind of thing gets caught by test failure soon.

Very worthwhile!

Reply:
Right, I was referring to the initial projection. The aggregate node, basically does:

```
kstream.mapValues(<all fields needed for aggregation>) // 1
    .selectKey() // 2
    .groupBy() // 3
    .aggregate()  // 4
    .filter()  // 5
    .mapValues(<final selection>)  // 6
```

Its that first projection (1) where we're potentially changing the order now, which is problematic because after an upgrade the consumer of the repartition topic might misinterpret the data (if the format is sensitive to field order) and the schema (if the field name changes). I think a similar concern applies to the data in the state store if the internal field order/names change.

The internal schema is determined here:

```
    InternalSchema(
        final Set<DereferenceExpression> requiredColumns,
        final List<Expression> aggregateFunctionArguments) {
      final Set<String> seen = new HashSet<>();
      collectAggregateArgExpressions(requiredColumns, seen);
      collectAggregateArgExpressions(aggregateFunctionArguments, seen);
    }
    private void collectAggregateArgExpressions(
        final Collection<? extends Expression> expressions,
        final Set<String> seen
    ) {
      expressions.stream()
          .filter(e -> !seen.contains(e.toString()))
          .forEach(expression -> {
            final String internalColumnName = INTERNAL_COLUMN_NAME_PREFIX
                + aggArgExpansionList.size();
            seen.add(expression.toString());
            internalNameToIndexMap.put(internalColumnName, aggArgExpansionList.size());
            aggArgExpansionList.add(SelectExpression.of(internalColumnName, expression));
            expressionToInternalColumnNameMap
                .putIfAbsent(expression.toString(), internalColumnName);
          });
    }
```
requiredColumns should have the same iteration order every time.

Reply:
Yeah, agreed - see comment below on the same.